storage_proxy: stop: await_pending_writes

bhalevy opened the pull request:
We take care to start_write() in the write response handlers,
but we don't await them on stop().  await_pending_writes() is
currently only used by storage_service::sstable_cleanup_fiber
to wait for all local writes to complete before cleanup.

await_pending_writes() also in stop() to make sure there are no
pending writes hanging around after storage_proxy is stopped.

A follow-up change will add utils::phased_barrier::close() to prevent new operations from starting after the service is stopped.

* Although this does not fix a known issue, I recommend backporting it to all live version, to be on the safe side.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14017/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2454/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5571/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13584/)
:x: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/237/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/835/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14017/)
#### Build Details:
- Duration: 3 hr 59 min
- Builder: spider8.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14017/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2454/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5571/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13584/) ❌ - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/237/)

https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/237/consoleFull:
```
12:35:49  curl: (28) Failed to connect to 169.254.169.254 port 80 after 10002 ms: Timeout was reached
```

Looks like random environmental flakiness

> ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/835/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14017/)
> 
> #### Build Details:
> * Duration: 3 hr 59 min
> * Builder: spider8.cloudius-systems.com

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -6931,7 +6931,7 @@ future<> storage_proxy::abort_view_writes() {
 future<>
 storage_proxy::stop() {
-    return make_ready_future<>();
+    return await_pending_writes();
```
Without aborting long running writes (currently MV timeout is 5 minutes or so) this may hang shutdown for a long time. For cleanup fiber this is less of a problem.

bhalevy replied:
I agree we should abort pending writes rather than wait for their timeout
but that said, I don't see we abort writes in table::stop(), so this is a pre-existing issue,
so I don't think it should block this PR.

gleb-cloudius replied:
Do we support MV for tablets already?

bhalevy replied:
> Do we support MV for tablets already?

No, not in 2025.1.0.
But how is this specific to tablets?

gleb-cloudius replied:
It is specific to MVs. Regular writes have timeout of 2 seconds. MVs have timeouts of 5 minutes IIRC.

bhalevy commented:
Converted to draft until https://github.com/scylladb/seastar/pull/2601

bhalevy commented:
> Converted to draft until [scylladb/seastar#2601](https://github.com/scylladb/seastar/pull/2601)

Part of https://github.com/scylladb/scylladb/pull/22076

bhalevy commented:
[1e0c855](https://github.com/scylladb/scylladb/commit/1e0c855c0a464cb5012edd6d97c40f6b518c2a12): rebased

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15130/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15130/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/cell_locker_test 🔹 boost/mutation_test 🔹 boost/view_build_test ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3315/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14449/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6432/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15130/)
> 
> #### Failed Tests (4/38839):
> * [test_lwt_truncate](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15130/testReport/junit/lwt_schema_modification_test/TestLWTSchemaModification/Tests___dtest___test_lwt_truncate) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_lwt_truncate)
> * [test_lwt_truncate](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15130/testReport/junit/lwt_schema_modification_test/TestLWTSchemaModification/Tests___dtest___test_lwt_truncate_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_lwt_truncate)
> * [test_lwt_truncate](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15130/testReport/junit/lwt_schema_modification_test/TestLWTSchemaModification/Tests___dtest_with_topology_changes___test_lwt_truncate) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_lwt_truncate)
> * [test_lwt_truncate](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15130/testReport/junit/lwt_schema_modification_test/TestLWTSchemaModification/Tests___dtest_with_topology_changes___test_lwt_truncate_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_lwt_truncate)

https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15130/artifact/dtest/logs-full.release.005/1739195178580_lwt_schema_modification_test.py%3A%3ATestLWTSchemaModification%3A%3Atest_lwt_truncate/node7.log
```
INFO  2025-02-10 13:45:12,207 [shard 0:main] init - Scylla version 2025.2.0~dev-0.20250210.3d68a84b3f7f with build-id d5a1483703eda6c1c474b9582d917ed91dd2a38b starting ...
...
scylla: seastar/include/seastar/core/rwlock.hh:115: void seastar::basic_rwlock<seastar::lowres_clock>::read_unlock() [Clock = seastar::lowres_clock]: Assertion `_sem.current() < max_ops' failed.
Aborting on shard 5, in scheduling group sl:default.
Backtrace:
  0xa91f70b
  0x1d94a5e
  0xa942d48
  /jenkins/workspace/scylla-master/gating-dtest-release/scylla/.ccm/scylla-repository/1e0c855c0a464cb5012edd6d97c40f6b518c2a12/libreloc/libc.so.6+0x1a04f
  /jenkins/workspace/scylla-master/gating-dtest-release/scylla/.ccm/scylla-repository/1e0c855c0a464cb5012edd6d97c40f6b518c2a12/libreloc/libc.so.6+0x73113
  /jenkins/workspace/scylla-master/gating-dtest-release/scylla/.ccm/scylla-repository/1e0c855c0a464cb5012edd6d97c40f6b518c2a12/libreloc/libc.so.6+0x19f9d
  /jenkins/workspace/scylla-master/gating-dtest-release/scylla/.ccm/scylla-repository/1e0c855c0a464cb5012edd6d97c40f6b518c2a12/libreloc/libc.so.6+0x1941
  /jenkins/workspace/scylla-master/gating-dtest-release/scylla/.ccm/scylla-repository/1e0c855c0a464cb5012edd6d97c40f6b518c2a12/libreloc/libc.so.6+0x185d
  /jenkins/workspace/scylla-master/gating-dtest-release/scylla/.ccm/scylla-repository/1e0c855c0a464cb5012edd6d97c40f6b518c2a12/libreloc/libc.so.6+0x120c6
  0x1923f54
  0x443ac86
  0x1d94de6
  0x2f48347
  0xa943481
  0xa8d50ea
  /jenkins/workspace/scylla-master/gating-dtest-release/scylla/.ccm/scylla-repository/1e0c855c0a464cb5012edd6d97c40f6b518c2a12/libreloc/libc.so.6+0x71167
  /jenkins/workspace/scylla-master/gating-dtest-release/scylla/.ccm/scylla-repository/1e0c855c0a464cb5012edd6d97c40f6b518c2a12/libreloc/libc.so.6+0xf514b
```

Decoded:
```
[Backtrace #0]
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at ./build/release/seastar/./seastar/include/seastar/util/backtrace.hh:69
seastar::print_with_backtrace(seastar::backtrace_buffer&, bool) at ./build/release/seastar/./build/release/seastar/./seastar/src/core/reactor.cc:805
seastar::print_with_backtrace(char const*, bool) at ./build/release/seastar/./build/release/seastar/./seastar/src/core/reactor.cc:850
 (inlined by) seastar::sigabrt_action() at ./build/release/seastar/./build/release/seastar/./seastar/src/core/reactor.cc:3994
 (inlined by) seastar::install_oneshot_signal_handler<6, (void (*)())(&seastar::sigabrt_action)>()::{lambda(int, siginfo_t*, void*)#1}::operator()(int, siginfo_t*, void*) const at ./build/release/seastar/./build/release/seastar/./seastar/src/core/reactor.cc:3971
 (inlined by) seastar::install_oneshot_signal_handler<6, (void (*)())(&seastar::sigabrt_action)>()::{lambda(int, siginfo_t*, void*)#1}::__invoke(int, siginfo_t*, void*) at ./build/release/seastar/./build/release/seastar/./seastar/src/core/reactor.cc:3966
/data/scylla-s3-reloc.cache/by-build-id/d5a1483703eda6c1c474b9582d917ed91dd2a38b/extracted/scylla/libreloc/libc.so.6: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=515c33a35f41020661fea8ac4eb995e26ccd6b00, for GNU/Linux 3.2.0, not stripped

__GI___sigaction at :?
__pthread_kill_implementation at ??:?
__GI_raise at :?
__GI_abort at :?
__assert_fail_base.cold at ??:?
__assert_fail at ??:?
seastar::basic_rwlock<seastar::lowres_clock>::read_unlock() at ././seastar/include/seastar/core/rwlock.hh:115
 (inlined by) row_locker::unlock(dht::decorated_key const*, bool, clustering_key_prefix const*, bool) at ././db/view/row_locking.cc:?
 (inlined by) row_locker::lock_holder::~lock_holder() at ././db/view/row_locking.cc:234
replica::database::do_apply(seastar::lw_shared_ptr<schema const>, frozen_mutation const&, tracing::trace_state_ptr, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, seastar::bool_class<db::force_sync_tag>, std::variant<std::monostate, db::per_partition_rate_limit::account_only, db::per_partition_rate_limit::account_and_enforce>) at ././replica/database.cc:2058
std::__n4861::coroutine_handle<seastar::internal::coroutine_traits_base<void>::promise_type>::resume() const at /usr/lib/gcc/x86_64-redhat-linux/14/../../../../include/c++/14/coroutine:242
 (inlined by) seastar::internal::coroutine_traits_base<void>::promise_type::run_and_dispose() at ./build/release/seastar/./seastar/include/seastar/core/coroutine.hh:122
 (inlined by) seastar::reactor::run_tasks(seastar::reactor::task_queue&) at ./build/release/seastar/./build/release/seastar/./seastar/src/core/reactor.cc:2606
 (inlined by) seastar::reactor::run_some_tasks() at ./build/release/seastar/./build/release/seastar/./seastar/src/core/reactor.cc:3078
seastar::reactor::do_run() at ./build/release/seastar/./build/release/seastar/./seastar/src/core/reactor.cc:3246
std::_Function_handler<void (), seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0>::_M_invoke(std::_Any_data const&) at ./build/release/seastar/./build/release/seastar/./seastar/src/core/reactor.cc:4555
seastar::posix_thread::start_routine(void*) at /usr/lib/gcc/x86_64-redhat-linux/14/../../../../include/c++/14/bits/std_function.h:591
```

Could be related to this change.

bhalevy commented:
In v3 ([b1e3c4b](https://github.com/scylladb/scylladb/commit/b1e3c4b9ec8c41554bcdcb7a63db59da1f923543)):
- fixed issue in the `row_locker: make lock_pk and lock_ck abortable` patch (that was caught by the test_let_truncate dtest)
- rebased
- converted storage_proxy::stop only as the last patch
  - after the patches to make the ops abortable
  - close() the phased barrier rather than just awaiting the pending ops

bhalevy commented:
In v4 ([6b3ac48](https://github.com/scylladb/scylladb/commit/6b3ac485f89e59aa008dec79ef7bb7f7c3e96e0c)):
- rebased
- utils: composite_abort_source: propagate exception to composite_abort_source

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15338/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15338/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cell_locker_test
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/view_build_test
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3470/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6589/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14608/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15338/)
#### Build Details:
- Duration: 5 hr 22 min
- Builder: spider4.cloudius-systems.com

bhalevy commented:
@scylladb/scylla-maint please consider merging

denesb commented:
> Although this does not fix a known issue, I recommend backporting it to all live version, to be on the safe side.

Need issue if you want this to be backported.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -6932,7 +6932,7 @@ future<> storage_proxy::abort_view_writes() {
 future<>
 storage_proxy::stop() {
-    return make_ready_future<>();
+    return _pending_writes_phaser.close();
```
I am not sure what difference this makes since IIRC this function is never called.

bhalevy replied:
It's called in cql_test_env, but indeed not called in main.
I'm actually testing right a now a patch to call proxy.stop() in main as well.

bhalevy replied:
Added to v5 ([9afde52](https://github.com/scylladb/scylladb/commit/9afde5277be7ef88dfab45e07026ca39872909b5))

gleb-cloudius replied:
>  I'm actually testing right a now a patch to call proxy.stop() in main as well.

Here be dragons

bhalevy replied:
Added an async gate to query processor held around queries and closed on stop
plus added a call to qp.stop in main - this prevents use of storage proxy via the query processor after the proxy is stopped (and shards destroyed).

bhalevy commented:
In v5 ([9afde52](https://github.com/scylladb/scylladb/commit/9afde5277be7ef88dfab45e07026ca39872909b5)):
- rebased
- main: stop storage_proxy

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15388/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15388/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cell_locker_test
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/view_build_test
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3509/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14647/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6628/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15388/)
#### Failed Tests (3/38815):
- [test_topology_streaming_failure.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15388/testReport/junit/topology/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure.debug.3)
- [test_decommissioned_node_cant_rejoin.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15388/testReport/junit/topology_custom/test_decommission/Tests___Unit_Tests___test_decommissioned_node_cant_rejoin_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommissioned_node_cant_rejoin.release.2)
- [test_decommissioned_node_cant_rejoin.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15388/testReport/junit/topology_custom/test_decommission/Tests___Unit_Tests___test_decommissioned_node_cant_rejoin_release_2_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommissioned_node_cant_rejoin.release.2)
#### Build Details:
- Duration: 5 hr 21 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
In v6 ([88acb98](https://github.com/scylladb/scylladb/commit/88acb9893aedebf1b7e3985fbab46c7f5fb9a2c8)):
- rebased
- query_processor: add async gate
- transport: server: hold the query processor gate

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15413/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15413/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cell_locker_test
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/view_build_test
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3530/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6651/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14669/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15413/)
#### Failed Tests (6/38860):
- [test_removenode_rejoin](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15413/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_removenode_rejoin) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_removenode_rejoin)
- [test_removenode_rejoin](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15413/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_removenode_rejoin_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_removenode_rejoin)
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15413/testReport/junit/boost.encryption_at_rest_test.test_kms_network_error.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [test_topology_streaming_failure.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15413/testReport/junit/topology/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure.debug.1)
- [test_topology_streaming_failure.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15413/testReport/junit/topology/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure.debug.2)
- [test_topology_streaming_failure.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15413/testReport/junit/topology/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure.debug.3)
#### Build Details:
- Duration: 8 hr 40 min
- Builder: spider6.cloudius-systems.com

bhalevy commented:
In v7 ([d0290e7](https://github.com/scylladb/scylladb/commit/d0290e716e4ad28d0af0b7002b7641722fc7169d)):
- rebased
- query_processor: add pending_operations phased barrier
- service_permit: hold phased_barrier operation
- storage_proxy: stop: close _pending_writes_phaser

The above allows gracefully stopping both the query_processor, then the storage_proxy

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15447/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15447/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cell_locker_test
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/view_build_test
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6675/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3553/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14693/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15447/)
#### Failed Tests (4/38894):
- [test_removenode_rejoin](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15447/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_removenode_rejoin) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_removenode_rejoin)
- [test_removenode_rejoin](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15447/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_removenode_rejoin_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_removenode_rejoin)
- [test_topology_streaming_failure.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15447/testReport/junit/topology/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure.debug.3)
- [test_tablet_repair_task_list.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15447/testReport/junit/topology_tasks/test_tablet_tasks/Tests___Unit_Tests___test_tablet_repair_task_list_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_repair_task_list.debug.3)
#### Build Details:
- Duration: 5 hr 37 min
- Builder: spider7.cloudius-systems.com

bhalevy commented:
In v8 ([2369667](https://github.com/scylladb/scylladb/commit/2369667fba2f46950a5127bff74f850c587d2582)):
- avoid empty_service_permit where possible by start an operation on the query processor or storage proxy kept
- Stop of query processor and storage proxy only after stopping database (i.e. the deferred actions are moved earlier in the code to be destroyed later)

gleb-cloudius commented:
>     * Stop of query processor and storage proxy only after stopping database (i.e. the deferred actions are moved earlier in the code to be destroyed later)

How does it make sense? query processor  uses storage proxy that uses database. That should be the order in which the are stopped.

bhalevy commented:
> > ```
> > * Stop of query processor and storage proxy only after stopping database (i.e. the deferred actions are moved earlier in the code to be destroyed later)
> > ```
> 
> How does it make sense? query processor uses storage proxy that uses database. That should be the order in which the are stopped.

Isn't the dependency circular?

gleb-cloudius commented:
> > > ```
> > > * Stop of query processor and storage proxy only after stopping database (i.e. the deferred actions are moved earlier in the code to be destroyed later)
> > > ```
> > 
> > 
> > How does it make sense? query processor uses storage proxy that uses database. That should be the order in which the are stopped.
> 
> Isn't the dependency circular?

What database needs storage proxy for? But even if it is circular jangling stopping order will not make it non circular.

bhalevy commented:
> > > > ```
> > > > * Stop of query processor and storage proxy only after stopping database (i.e. the deferred actions are moved earlier in the code to be destroyed later)
> > > > ```
> > > 
> > > 
> > > How does it make sense? query processor uses storage proxy that uses database. That should be the order in which the are stopped.
> > 
> > 
> > Isn't the dependency circular?
> 
> What database needs storage proxy for? But even if it is circular jangling stopping order will not make it non circular.

Via compaction and large data handler updating tables (compaction history, large data tables).
Would stopping remote (on both sides of the wire) before local proxy help?

gleb-cloudius commented:
> > > > > ```
> > > > > * Stop of query processor and storage proxy only after stopping database (i.e. the deferred actions are moved earlier in the code to be destroyed later)
> > > > > ```
> > > > 
> > > > 
> > > > How does it make sense? query processor uses storage proxy that uses database. That should be the order in which the are stopped.
> > > 
> > > 
> > > Isn't the dependency circular?
> > 
> > 
> > What database needs storage proxy for? But even if it is circular jangling stopping order will not make it non circular.
> 
> Via compaction and large data handler updating tables (compaction history, large data tables). Would stopping remote (on both sides of the wire on the stopping node) before local proxy help?

Compactions have their own module - compaction_manager. Why not stopping that before qp and sp? not sure about large data handler. At which point those are happening?

avikivity commented:
> > > > ```
> > > > * Stop of query processor and storage proxy only after stopping database (i.e. the deferred actions are moved earlier in the code to be destroyed later)
> > > > ```
> > > 
> > > 
> > > How does it make sense? query processor uses storage proxy that uses database. That should be the order in which the are stopped.
> > 
> > 
> > Isn't the dependency circular?
> 
> What database needs storage proxy for? But even if it is circular jangling stopping order will not make it non circular.

database -> view update -> storage_proxy -> database

avikivity commented:
Please annotate the cover letter with perf-simple-query results.

The cover letter should also address the circular dependencies. Lots of details in the patches (I presume), but nothing on the big picture.

avikivity commented:
> > ```
> > * Stop of query processor and storage proxy only after stopping database (i.e. the deferred actions are moved earlier in the code to be destroyed later)
> > ```
> 
> How does it make sense? query processor uses storage proxy that uses database. That should be the order in which the are stopped.

Correct but incomplete (due to the circular dependency). The order @gleb-cloudius lists is the main order.

avikivity commented:
> 
> * Depends on [Phased barrier improvements #22018](https://github.com/scylladb/scylladb/pull/22018), so can be backported as far back as branch-2025.1

Nobody will backport such a huge series, without very strong motivation.

gleb-cloudius commented:
> > > > > ```
> > > > > * Stop of query processor and storage proxy only after stopping database (i.e. the deferred actions are moved earlier in the code to be destroyed later)
> > > > > ```
> > > > 
> > > > 
> > > > How does it make sense? query processor uses storage proxy that uses database. That should be the order in which the are stopped.
> > > 
> > > 
> > > Isn't the dependency circular?
> > 
> > 
> > What database needs storage proxy for? But even if it is circular jangling stopping order will not make it non circular.
> 
> database -> view update -> storage_proxy -> database

As far as I see view updater is a separate service. So shouldn't this be broken by making view update depend on the database (by means of subscribing for changes for instance) and then view update  will be stopped before sp.

gleb-cloudius commented:
> > > > > > > ```
> > > > > > > * Stop of query processor and storage proxy only after stopping database (i.e. the deferred actions are moved earlier in the code to be destroyed later)
> > > > > > > ```
> > > > > > 
> > > > > > 
> > > > > > How does it make sense? query processor uses storage proxy that uses database. That should be the order in which the are stopped.
> > > > > 
> > > > > 
> > > > > Isn't the dependency circular?
> > > > 
> > > > 
> > > > What database needs storage proxy for? But even if it is circular jangling stopping order will not make it non circular.
> > > 
> > > 
> > > Via compaction and large data handler updating tables (compaction history, large data tables). Would stopping remote (on both sides of the wire on the stopping node) before local proxy help?
> > 
> > 
> > Compactions have their own module - compaction_manager. Why not stopping that before qp and sp?
> 
> compaction_manager is at a low level, it is started long before storage_proxy (it's part of the replica).
> 

I see. But than what does it do while storage_proxy is not available?

bhalevy commented:
> > * Depends on [Phased barrier improvements #22018](https://github.com/scylladb/scylladb/pull/22018), so can be backported as far back as branch-2025.1
> 
> Nobody will backport such a huge series, without very strong motivation.

Yes, agreed. This was left over from the early days of this patch when it was small and innocent.

bhalevy commented:
> > > > > > > > ```
> > > > > > > > * Stop of query processor and storage proxy only after stopping database (i.e. the deferred actions are moved earlier in the code to be destroyed later)
> > > > > > > > ```
> > > > > > > 
> > > > > > > 
> > > > > > > How does it make sense? query processor uses storage proxy that uses database. That should be the order in which the are stopped.
> > > > > > 
> > > > > > 
> > > > > > Isn't the dependency circular?
> > > > > 
> > > > > 
> > > > > What database needs storage proxy for? But even if it is circular jangling stopping order will not make it non circular.
> > > > 
> > > > 
> > > > Via compaction and large data handler updating tables (compaction history, large data tables). Would stopping remote (on both sides of the wire on the stopping node) before local proxy help?
> > > 
> > > 
> > > Compactions have their own module - compaction_manager. Why not stopping that before qp and sp?
> > 
> > 
> > compaction_manager is at a low level, it is started long before storage_proxy (it's part of the replica).
> 
> I see. But than what does it do while storage_proxy is not available?

It doesn't know that.

If we had a registration / deregistration entry point we could deregister the query_processor / storage_proxy before stopping them.  Then, the compaction_manager (for example) can skip updating compaction history.  Similarly the large data handler can skip writing to the local large data system tables.

That said, it would be better to drain the system top-down as Avi [suggested](https://github.com/scylladb/scylladb/pull/22017#issuecomment-2676831062), and then when the system is shut down and nothing should be in flight, stopping and destroying the services would be easier.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15516/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15516/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cell_locker_test
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/view_build_test
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3607/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14750/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6727/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15516/)
#### Failed Tests (7/38800):
- [test_only_superuser_alters_superuser_status](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15516/testReport/junit/auth_test/TestAuth/Tests___dtest___test_only_superuser_alters_superuser_status) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_only_superuser_alters_superuser_status)
- [test_only_superuser_alters_superuser_status](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15516/testReport/junit/auth_test/TestAuth/Tests___dtest___test_only_superuser_alters_superuser_status_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_only_superuser_alters_superuser_status)
- [test_role_requires_login_privilege_to_authenticate](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15516/testReport/junit/auth_roles_test/TestAuthRoles/Tests___dtest___test_role_requires_login_privilege_to_authenticate) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_role_requires_login_privilege_to_authenticate)
- [test_role_requires_login_privilege_to_authenticate](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15516/testReport/junit/auth_roles_test/TestAuthRoles/Tests___dtest___test_role_requires_login_privilege_to_authenticate_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_role_requires_login_privilege_to_authenticate)
- [test_role_admin_validation](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15516/testReport/junit/auth_roles_test/TestAuthRoles/Tests___dtest___test_role_admin_validation) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_role_admin_validation)
- [test_role_admin_validation](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15516/testReport/junit/auth_roles_test/TestAuthRoles/Tests___dtest___test_role_admin_validation_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_role_admin_validation)
- [test_topology_ops_encrypted[False].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15516/testReport/junit/topology_custom/test_topology_ops_encrypted/Tests___Unit_Tests___test_topology_ops_encrypted_False__debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_ops_encrypted[False].debug.1)
#### Build Details:
- Duration: 5 hr 28 min
- Builder: spider8.cloudius-systems.com

avikivity commented:
> > > > > > ```
> > > > > > * Stop of query processor and storage proxy only after stopping database (i.e. the deferred actions are moved earlier in the code to be destroyed later)
> > > > > > ```
> > > > > 
> > > > > 
> > > > > How does it make sense? query processor uses storage proxy that uses database. That should be the order in which the are stopped.
> > > > 
> > > > 
> > > > Isn't the dependency circular?
> > > 
> > > 
> > > What database needs storage proxy for? But even if it is circular jangling stopping order will not make it non circular.
> > 
> > 
> > database -> view update -> storage_proxy -> database
> 
> As far as I see view updater is a separate service. 

View updates are generated on the write path of replica::database.


> So shouldn't this be broken by making view update depend on the database (by means of subscribing for changes for instance) and then view update will be stopped before sp.

It's still circular. replica::database -> view updates -> storage_proxy -> replica::database.

mykaul commented:
@bhalevy - please resolve conflicts, if this is still relevant (or better yet, delegate this PR to someone)

bhalevy commented:
> @bhalevy - please resolve conflicts, if this is still relevant (or better yet, delegate this PR to someone)

@piodul it'd be best if your team will take over this PR

emdroid commented:
Yes, we'll continue on this one 👍

bhalevy commented:
- rebased and fixed seastar includes to use angle brackets

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17475/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17475/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17475/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cell_locker_test
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/view_build_test
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8527/)
:x: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16611/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5360/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17475/)
#### Failed Tests (13/45554):
- [test_role_admin_validation](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17475/testReport/junit/auth_roles_test/TestAuthRoles/Tests___dtest_with_gossip_topology_changes___test_role_admin_validation) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_role_admin_validation)
- [test_role_admin_validation](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17475/testReport/junit/auth_roles_test/TestAuthRoles/Tests___dtest_with_gossip_topology_changes___test_role_admin_validation_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_role_admin_validation)
- [test_role_requires_login_privilege_to_authenticate](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17475/testReport/junit/auth_roles_test/TestAuthRoles/Tests___dtest_with_gossip_topology_changes___test_role_requires_login_privilege_to_authenticate) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_role_requires_login_privilege_to_authenticate)
- [test_role_requires_login_privilege_to_authenticate](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17475/testReport/junit/auth_roles_test/TestAuthRoles/Tests___dtest_with_gossip_topology_changes___test_role_requires_login_privilege_to_authenticate_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_role_requires_login_privilege_to_authenticate)
- [test_only_superuser_alters_superuser_status](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17475/testReport/junit/auth_test/TestAuth/Tests___dtest_with_gossip_topology_changes___test_only_superuser_alters_superuser_status) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_only_superuser_alters_superuser_status)
- [test_only_superuser_alters_superuser_status](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17475/testReport/junit/auth_test/TestAuth/Tests___dtest_with_gossip_topology_changes___test_only_superuser_alters_superuser_status_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_only_superuser_alters_superuser_status)
- [test_raft_recovery_user_data[remove].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17475/testReport/junit/cluster/test_raft_recovery_user_data/Tests___Unit_Tests___test_raft_recovery_user_data_remove__debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_raft_recovery_user_data[remove].debug.1)
- [test_raft_recovery_user_data[remove].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17475/testReport/junit/cluster/test_raft_recovery_user_data/Tests___Unit_Tests___test_raft_recovery_user_data_remove__debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_raft_recovery_user_data[remove].debug.2)
- [test_raft_recovery_user_data[remove].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17475/testReport/junit/cluster/test_raft_recovery_user_data/Tests___Unit_Tests___test_raft_recovery_user_data_remove__debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_raft_recovery_user_data[remove].debug.3)
- [test_raft_recovery_user_data[replace].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17475/testReport/junit/cluster/test_raft_recovery_user_data/Tests___Unit_Tests___test_raft_recovery_user_data_replace__debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_raft_recovery_user_data[replace].debug.1)
#### Build Details:
- Duration: 6 hr 54 min
- Builder: spider5.cloudius-systems.com
